File: nodejs/src/validation.ts
import { FhirPatient, OpenEhrComposition, ValidationIssue } from './models';

/** Validates that a string is well-formed FHIR JSON for the expected resource type. */
export function validateFhirPatientJson(json: string): {
  issues: ValidationIssue[];
  patient?: FhirPatient;
} {
  const issues: ValidationIssue[] = [];

  if (!json || json.trim() === '') {
    issues.push({ severity: 'error', message: 'FHIR payload is empty.' });
    return { issues };
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(json);
  } catch (err) {
    issues.push({
      severity: 'error',
      message: `Payload is not valid FHIR JSON: ${(err as Error).message}`,
    });
    return { issues };
  }

  if (typeof parsed !== 'object' || parsed === null) {
    issues.push({ severity: 'error', message: 'Payload did not parse to a FHIR resource.' });
    return { issues };
  }

  const resource = parsed as FhirPatient;
  if (resource.resourceType !== 'Patient') {
    issues.push({
      severity: 'error',
      message: `Expected a FHIR 'Patient' resource but received '${resource.resourceType}'.`,
      location: 'resourceType',
    });
    return { issues };
  }

  return { issues, patient: resource };
}

/** Validates the structure of an openEHR composition before mapping. */
export function validateComposition(composition: OpenEhrComposition | null | undefined): {
  issues: ValidationIssue[];
} {
  const issues: ValidationIssue[] = [];

  if (!composition) {
    issues.push({ severity: 'error', message: 'openEHR composition is null.' });
    return { issues };
  }

  if (!composition.archetypeNodeId) {
    issues.push({
      severity: 'error',
      message: "Composition is missing 'archetypeNodeId'.",
      location: 'archetypeNodeId',
    });
  }

  if (!composition.demographics) {
    issues.push({
      severity: 'error',
      message: 'Composition does not contain a demographics payload.',
      location: 'demographics',
    });
  } else if (!composition.demographics.familyName && !composition.demographics.givenName) {
    issues.push({
      severity: 'warning',
      message: 'Demographics contain neither a family name nor a given name.',
      location: 'demographics.familyName',
    });
  }

  if (!composition.ehrStatus || !composition.ehrStatus.subjectId) {
    issues.push({
      severity: 'warning',
      message:
        'EHR_STATUS has no subject id; the produced FHIR Patient will be assigned a generated id.',
      location: 'ehrStatus.subjectId',
    });
  }

  return { issues };
}

File: rust/src/validation.rs
//! Inbound payload validation.

use crate::models::*;
use serde_json::Value;

/// Parses and validates inbound FHIR Patient JSON, returning issues and (on
/// success) the parsed JSON value.
pub fn validate_fhir_patient_json(json: &str) -> (Vec<ValidationIssue>, Option<Value>) {
    let mut issues = Vec::new();

    if json.trim().is_empty() {
        issues.push(ValidationIssue::error("FHIR payload is empty.", None));
        return (issues, None);
    }

    let parsed: Value = match serde_json::from_str(json) {
        Ok(v) => v,
        Err(e) => {
            issues.push(ValidationIssue::error(
                format!("Payload is not valid FHIR JSON: {e}"),
                None,
            ));
            return (issues, None);
        }
    };

    let resource_type = parsed.get("resourceType").and_then(|v| v.as_str());
    if resource_type != Some("Patient") {
        issues.push(ValidationIssue::error(
            format!(
                "Expected a FHIR 'Patient' resource but received '{}'.",
                resource_type.unwrap_or("")
            ),
            Some("resourceType"),
        ));
        return (issues, None);
    }

    (issues, Some(parsed))
}

/// Validates an openEHR composition before mapping.
pub fn validate_composition(c: &OpenEhrComposition) -> Vec<ValidationIssue> {
    let mut issues = Vec::new();

    if c.archetype_node_id.trim().is_empty() {
        issues.push(ValidationIssue::error(
            "Composition is missing 'archetypeNodeId'.",
            Some("archetypeNodeId"),
        ));
    }

    match &c.demographics {
        None => issues.push(ValidationIssue::error(
            "Composition does not contain a demographics payload.",
            Some("demographics"),
        )),
        Some(d) => {
            let no_family = d.family_name.as_deref().unwrap_or("").is_empty();
            let no_given = d.given_name.as_deref().unwrap_or("").is_empty();
            if no_family && no_given {
                issues.push(ValidationIssue::warning(
                    "Demographics contain neither a family name nor a given name.",
                    Some("demographics.familyName"),
                ));
            }
        }
    }

    if c.ehr_status.subject_id.as_deref().unwrap_or("").trim().is_empty() {
        issues.push(ValidationIssue::warning(
            "EHR_STATUS has no subject id; the produced FHIR Patient will be assigned a generated id.",
            Some("ehrStatus.subjectId"),
        ));
    }

    issues
}
